# RQ2. Forecasting Transaction Value and Active Accounts

# Problem Statement
Can machine learning forecast future mobile money transaction value and active-account growth on a short national monthly panel (84 months)?
# Business Context
Operators and the Bank of Ghana need dependable monthly forecasts for liquidity, agent placement, and payment-stability monitoring.
# Objectives
Test stationarity (ADF, KPSS); fit SARIMA via auto_arima, Prophet with an E-Levy regressor, and XGBoost and LightGBM on lag features tuned with a TimeSeriesSplit; and evaluate on the 2025 12-month holdout for both targets.


In [1]:
# --- Colab setup: install dependencies and load data (run once) ---
import sys, subprocess
def _pip(pkgs):
    subprocess.run([sys.executable, "-m", "pip", "install", "-q", *pkgs])
try:
    import pmdarima, prophet, shap, xgboost, lightgbm, imblearn  # noqa
except Exception:
    _pip(["pmdarima", "prophet", "shap", "xgboost", "lightgbm", "imbalanced-learn", "seaborn"])

import os
# Clone the repository if the processed data are not already present
if not os.path.exists("data/processed/monthly_series_clean.csv"):
    if not os.path.exists("mobile-money-ghana-forecasting"):
        subprocess.run(["git", "clone", "--depth", "1",
                        "https://github.com/bcudjoe/mobile-money-ghana-forecasting.git"])
    os.chdir("mobile-money-ghana-forecasting")

RANDOM_STATE = 42  # fixed seed for reproducibility
os.makedirs("outputs/figures", exist_ok=True)
os.makedirs("results", exist_ok=True)
print("Setup complete. Reproducible run, seed", RANDOM_STATE)


/usr/local/lib/python3.11/dist-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
Importing plotly failed. Interactive plots will not work.


Setup complete. Reproducible run, seed 42


## Analysis
The code below is the exact, tested pipeline that produces the figures in `outputs/figures/` and the metrics in `results/`. It runs top to bottom on a fresh Colab runtime.

In [2]:
import json, warnings
import numpy as np
import pandas as pd
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
from statsmodels.tsa.stattools import adfuller, kpss
from statsmodels.stats.diagnostic import acorr_ljungbox
import pmdarima as pm
from prophet import Prophet
import xgboost as xgb
import lightgbm as lgb
from sklearn.model_selection import TimeSeriesSplit, RandomizedSearchCV
from scipy.stats import uniform, randint

warnings.filterwarnings("ignore")
RANDOM_STATE = 42
np.random.seed(RANDOM_STATE)
FIG = "outputs/figures"
plt.rcParams.update({"figure.dpi": 200, "savefig.dpi": 200, "font.size": 12,
                     "axes.titlesize": 14, "axes.labelsize": 12})
BLUE, ORANGE, GREEN, GREY = "#2166AC", "#D6604D", "#1B7837", "#888888"

df = pd.read_csv("data/processed/monthly_series_clean.csv", parse_dates=["date"]).sort_values("date").reset_index(drop=True)
df["month"] = df["date"].dt.month

EXOG = ["agent_density", "mobile_pen", "internet_pen", "account_ownership",
        "inflation", "policy_rate", "exch_rate", "gdp_proxy", "elevy"]
TEST_START = pd.Timestamp("2025-01-01")

def metrics(y_true, y_pred):
    y_true, y_pred = np.asarray(y_true, float), np.asarray(y_pred, float)
    rmse = float(np.sqrt(np.mean((y_true - y_pred) ** 2)))
    mae = float(np.mean(np.abs(y_true - y_pred)))
    mape = float(np.mean(np.abs((y_true - y_pred) / y_true)) * 100)
    ss_res = np.sum((y_true - y_pred) ** 2)
    ss_tot = np.sum((y_true - np.mean(y_true)) ** 2)
    r2 = float(1 - ss_res / ss_tot)
    return {"RMSE": round(rmse, 1), "MAE": round(mae, 1), "MAPE": round(mape, 2), "R2": round(r2, 3)}

def run_target(target, tag):
    out = {"target": target}
    s = df.set_index("date")[target].astype(float)
    train = s[s.index < TEST_START]
    test = s[s.index >= TEST_START]
    out["n_train"], out["n_test"] = len(train), len(test)

    # --- Stationarity diagnostics ---
    adf_p = adfuller(train.dropna())[1]
    try:
        kpss_p = kpss(train.dropna(), nlags="auto")[1]
    except Exception:
        kpss_p = None
    adf_p_d1 = adfuller(train.diff().dropna())[1]
    out["stationarity"] = {"adf_p_level": round(float(adf_p), 4),
                           "kpss_p_level": (round(float(kpss_p), 4) if kpss_p is not None else None),
                           "adf_p_diff1": round(float(adf_p_d1), 4)}

    preds = {}

    # --- Seasonal-naive benchmark (value 12 months earlier) ---
    snaive = s.shift(12).reindex(test.index)
    preds["Seasonal-naive"] = snaive.values

    # --- SARIMA via auto_arima (search grid documented) ---
    sar = pm.auto_arima(train, seasonal=True, m=12, d=None, D=1,
                        start_p=0, max_p=3, start_q=0, max_q=3,
                        max_P=2, max_Q=2, information_criterion="aic",
                        stepwise=True, suppress_warnings=True, error_action="ignore",
                        random_state=RANDOM_STATE)
    sar_fc = sar.predict(n_periods=len(test))
    preds["SARIMA"] = np.asarray(sar_fc)
    out["sarima_order"] = {"order": list(sar.order), "seasonal_order": list(sar.seasonal_order),
                           "aic": round(float(sar.aic()), 1)}
    # Ljung-Box on SARIMA residuals
    lb = acorr_ljungbox(sar.resid(), lags=[12], return_df=True)
    out["sarima_ljungbox_p_lag12"] = round(float(lb["lb_pvalue"].iloc[0]), 4)

    # --- Prophet with E-Levy regressor + yearly seasonality ---
    pdf = df[df["date"] < TEST_START][["date", target, "elevy"]].rename(columns={"date": "ds", target: "y"})
    fdf = df[["date", "elevy"]].rename(columns={"date": "ds"})
    best_ph, best_mape = None, np.inf
    grid = [(cp, sp, mode) for cp in [0.05, 0.1, 0.5] for sp in [10.0] for mode in ["additive", "multiplicative"]]
    # validate on last 12 of train
    ptr = pdf[pdf["ds"] < pd.Timestamp("2024-01-01")]
    pval_idx = pdf[pdf["ds"] >= pd.Timestamp("2024-01-01")]
    for cp, sp, mode in grid:
        try:
            mm = Prophet(yearly_seasonality=True, weekly_seasonality=False, daily_seasonality=False,
                         changepoint_prior_scale=cp, seasonality_prior_scale=sp, seasonality_mode=mode)
            mm.add_regressor("elevy")
            mm.fit(ptr)
            fut = pval_idx[["ds", "elevy"]]
            fc = mm.predict(fut)["yhat"].values
            mp = metrics(pval_idx["y"].values, fc)["MAPE"]
            if mp < best_mape:
                best_mape, best_ph = mp, (cp, sp, mode)
        except Exception:
            continue
    cp, sp, mode = best_ph
    ph = Prophet(yearly_seasonality=True, weekly_seasonality=False, daily_seasonality=False,
                 changepoint_prior_scale=cp, seasonality_prior_scale=sp, seasonality_mode=mode)
    ph.add_regressor("elevy")
    ph.fit(pdf)
    ph_fc = ph.predict(fdf[fdf["ds"] >= TEST_START][["ds", "elevy"]])["yhat"].values
    preds["Prophet"] = ph_fc
    out["prophet_params"] = {"changepoint_prior_scale": cp, "seasonality_prior_scale": sp,
                             "seasonality_mode": mode, "val_mape": round(best_mape, 2)}

    # --- Gradient boosting on lag features (recursive multi-step) ---
    lag_cols = [f"{target}_lag{L}" for L in range(1, 13)]
    feat = lag_cols + ["month", "t_index"] + EXOG
    d = df.copy()
    # rolling means of target
    d[f"{target}_roll3"] = d[target].rolling(3).mean()
    d[f"{target}_roll6"] = d[target].rolling(6).mean()
    feat = feat + [f"{target}_roll3", f"{target}_roll6"]
    model_ready = d.dropna(subset=lag_cols).reset_index(drop=True)
    tr = model_ready[model_ready["date"] < TEST_START]
    Xtr, ytr = tr[feat].values, tr[target].values

    tscv = TimeSeriesSplit(n_splits=5)
    best_params = {}
    for name, est, space in [
        ("XGBoost", xgb.XGBRegressor(random_state=RANDOM_STATE, n_jobs=-1, objective="reg:squarederror"),
         {"n_estimators": randint(200, 700), "max_depth": randint(2, 5),
          "learning_rate": uniform(0.01, 0.15), "subsample": uniform(0.7, 0.3),
          "colsample_bytree": uniform(0.7, 0.3), "min_child_weight": randint(1, 6),
          "reg_alpha": uniform(0, 1.0), "reg_lambda": uniform(0.5, 2.0)}),
        ("LightGBM", lgb.LGBMRegressor(random_state=RANDOM_STATE, n_jobs=-1, verbose=-1),
         {"n_estimators": randint(200, 700), "max_depth": randint(2, 6),
          "learning_rate": uniform(0.01, 0.15), "subsample": uniform(0.7, 0.3),
          "colsample_bytree": uniform(0.7, 0.3), "num_leaves": randint(8, 40),
          "min_child_samples": randint(5, 25), "reg_lambda": uniform(0.5, 2.0)}),
    ]:
        search = RandomizedSearchCV(est, space, n_iter=40, cv=tscv,
                                    scoring="neg_root_mean_squared_error",
                                    random_state=RANDOM_STATE, n_jobs=-1)
        search.fit(Xtr, ytr)
        best = search.best_estimator_
        best_params[name] = {k: (round(v, 4) if isinstance(v, float) else int(v))
                             for k, v in search.best_params_.items()}
        # recursive multi-step forecast over the test horizon
        hist = df.set_index("date")[target].astype(float).copy()
        exog_by_date = df.set_index("date")[EXOG + ["t_index"]]
        fc = []
        for dt in test.index:
            row = {}
            for L in range(1, 13):
                row[f"{target}_lag{L}"] = hist.get(dt - pd.DateOffset(months=L), np.nan)
            recent = hist[hist.index < dt]
            row[f"{target}_roll3"] = recent.tail(3).mean()
            row[f"{target}_roll6"] = recent.tail(6).mean()
            row["month"] = dt.month
            row["t_index"] = float(exog_by_date.loc[dt, "t_index"])
            for c in EXOG:
                row[c] = float(exog_by_date.loc[dt, c])
            xrow = np.array([[row[c] for c in feat]])
            yhat = float(best.predict(xrow)[0])
            fc.append(yhat)
            hist.loc[dt] = yhat  # feed prediction forward
        preds[name] = np.array(fc)
    out["gb_best_params"] = best_params

    # --- Metric table for the holdout ---
    table = {name: metrics(test.values, p) for name, p in preds.items()}
    out["holdout_metrics"] = table
    out["predictions"] = {name: [round(float(x), 1) for x in p] for name, p in preds.items()}
    out["actual"] = [round(float(x), 1) for x in test.values]
    out["test_dates"] = [d.strftime("%Y-%m") for d in test.index]

    # --- Plot forecasts vs actual ---
    plt.figure(figsize=(11, 6))
    hist_plot = s[s.index >= pd.Timestamp("2023-01-01")]
    plt.plot(hist_plot.index, hist_plot.values, color=GREY, lw=1.5, label="Actual (history)")
    plt.plot(test.index, test.values, color="black", lw=2.5, label="Actual (2025)")
    colors = {"SARIMA": BLUE, "Prophet": GREEN, "XGBoost": ORANGE, "LightGBM": "#9970AB",
              "Seasonal-naive": "#BBBBBB"}
    for name, p in preds.items():
        plt.plot(test.index, p, lw=1.8, ls="--", color=colors.get(name, None), label=name)
    plt.title(f"RQ2 forecast vs actual, {tag} (2025 holdout)")
    plt.ylabel(target); plt.legend(ncol=2, fontsize=9); plt.tight_layout()
    plt.savefig(f"{FIG}/rq2_forecast_{target}.png", bbox_inches="tight"); plt.close()

    best_model = min(table, key=lambda k: table[k]["RMSE"])
    out["best_model_holdout"] = best_model
    return out

results = {}
for target, tag in [("mm_value", "transaction value"), ("mm_active_accts", "active accounts")]:
    print(f"\n=== Running {target} ===")
    results[target] = run_target(target, tag)
    r = results[target]
    print("SARIMA order:", r["sarima_order"])
    print("Prophet params:", r["prophet_params"])
    for name, mt in r["holdout_metrics"].items():
        print(f"  {name:16s} RMSE={mt['RMSE']:>12} MAPE={mt['MAPE']:>6} R2={mt['R2']}")
    print("Best (holdout RMSE):", r["best_model_holdout"])

with open("results/rq2_results.json", "w") as fp:
    json.dump(results, fp, indent=2, default=str)
print("\nDONE")


=== Running mm_value ===


14:44:53 - cmdstanpy - INFO - Chain [1] start processing


14:44:53 - cmdstanpy - INFO - Chain [1] done processing


14:44:53 - cmdstanpy - INFO - Chain [1] start processing


14:44:53 - cmdstanpy - INFO - Chain [1] done processing


14:44:53 - cmdstanpy - INFO - Chain [1] start processing


14:44:53 - cmdstanpy - INFO - Chain [1] done processing


14:44:53 - cmdstanpy - INFO - Chain [1] start processing


14:44:54 - cmdstanpy - INFO - Chain [1] done processing


14:44:54 - cmdstanpy - INFO - Chain [1] start processing


14:44:54 - cmdstanpy - INFO - Chain [1] done processing


14:44:54 - cmdstanpy - INFO - Chain [1] start processing


14:44:54 - cmdstanpy - INFO - Chain [1] done processing


14:44:54 - cmdstanpy - INFO - Chain [1] start processing


14:44:55 - cmdstanpy - INFO - Chain [1] done processing


SARIMA order: {'order': [0, 1, 0], 'seasonal_order': [0, 1, 1, 12], 'aic': 1272.4}
Prophet params: {'changepoint_prior_scale': 0.1, 'seasonality_prior_scale': 10.0, 'seasonality_mode': 'additive', 'val_mape': 10.13}
  Seasonal-naive   RMSE=    131005.9 MAPE= 33.87 R2=-4.62
  SARIMA           RMSE=     37454.1 MAPE=   8.2 R2=0.541
  Prophet          RMSE=     35827.9 MAPE=  7.12 R2=0.58
  XGBoost          RMSE=     74673.6 MAPE= 12.69 R2=-0.826
  LightGBM         RMSE=     86776.3 MAPE= 16.88 R2=-1.466
Best (holdout RMSE): Prophet

=== Running mm_active_accts ===


14:45:09 - cmdstanpy - INFO - Chain [1] start processing


14:45:09 - cmdstanpy - INFO - Chain [1] done processing


14:45:09 - cmdstanpy - INFO - Chain [1] start processing


14:45:09 - cmdstanpy - INFO - Chain [1] done processing


14:45:09 - cmdstanpy - INFO - Chain [1] start processing


14:45:09 - cmdstanpy - INFO - Chain [1] done processing


14:45:09 - cmdstanpy - INFO - Chain [1] start processing


14:45:10 - cmdstanpy - INFO - Chain [1] done processing


14:45:10 - cmdstanpy - INFO - Chain [1] start processing


14:45:10 - cmdstanpy - INFO - Chain [1] done processing


14:45:10 - cmdstanpy - INFO - Chain [1] start processing


14:45:10 - cmdstanpy - INFO - Chain [1] done processing


14:45:10 - cmdstanpy - INFO - Chain [1] start processing


14:45:11 - cmdstanpy - INFO - Chain [1] done processing


SARIMA order: {'order': [0, 0, 1], 'seasonal_order': [0, 1, 0, 12], 'aic': 1815.5}
Prophet params: {'changepoint_prior_scale': 0.5, 'seasonality_prior_scale': 10.0, 'seasonality_mode': 'multiplicative', 'val_mape': 3.45}
  Seasonal-naive   RMSE=   1617612.1 MAPE=   5.1 R2=-2.591
  SARIMA           RMSE=   1188894.5 MAPE=  3.87 R2=-0.94
  Prophet          RMSE=    863704.0 MAPE=  2.86 R2=-0.024
  XGBoost          RMSE=   1359908.4 MAPE=  4.36 R2=-1.538
  LightGBM         RMSE=   2409940.2 MAPE=  6.59 R2=-6.97
Best (holdout RMSE): Prophet

DONE


## Observations on RQ2
- The value series is non-stationary in the level (ADF p about 1.0) and stationary after first differencing, so SARIMA uses one order of differencing.
- On the 12-month holdout, Prophet gives the lowest error for transaction value (MAPE about 7.1%), with SARIMA close (about 8.2%); the gradient-boosting models degrade over the long horizon because recursive forecasts compound errors on a short series.
- Every model beats the seasonal-naive benchmark (MAPE about 34%).
